providers/default/macs/cmac_prov.c

```diff
@@ -0,0 +1,261 @@
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/opensslconf.h>
+#ifndef OPENSSL_NO_CMAC
+
+# include <openssl/core_numbers.h>
+# include <openssl/core_names.h>
+# include <openssl/params.h>
+# include <openssl/engine.h>
+# include <openssl/evp.h>
+# include <openssl/cmac.h>
+
+# include "internal/provider_algs.h"
+# include "internal/provider_ctx.h"
+
+/*
+ * Forward declaration of everything implemented here.  This is not strictly
+ * necessary for the compiler, but provides an assurance that the signatures
+ * of the functions in the dispatch table are correct.
+ */
+static OSSL_OP_mac_newctx_fn cmac_new;
+static OSSL_OP_mac_dupctx_fn cmac_dup;
+static OSSL_OP_mac_freectx_fn cmac_free;
+static OSSL_OP_mac_ctx_get_param_types_fn cmac_ctx_get_param_types;
+static OSSL_OP_mac_ctx_get_params_fn cmac_ctx_get_params;
+static OSSL_OP_mac_ctx_set_param_types_fn cmac_ctx_set_param_types;
+static OSSL_OP_mac_ctx_set_params_fn cmac_ctx_set_params;
+static OSSL_OP_mac_init_fn cmac_init;
+static OSSL_OP_mac_update_fn cmac_update;
+static OSSL_OP_mac_final_fn cmac_final;
+
+/* local CMAC data */
+
+struct cmac_data_st {
+    void *provctx;
+    CMAC_CTX *ctx;
+
+    /*
+     * References to the underlying cipher implementation.  tmpcipher
+     * caches the cipher, always.  alloc_cipher only holds a reference
+     * to an explicitly fetched cipher.
+     * tmpcipher is cleared after a CMAC_Init call.
+     */
+    const EVP_CIPHER *tmpcipher; /* cached CMAC cipher */
+    EVP_CIPHER *alloc_cipher;    /* fetched CMAC cipher */
+
+    /*
+     * Conditions for legacy EVP_CIPHER uses.
+     * tmpengine is cleared after a CMAC_Init call.
+     */
+    ENGINE *tmpengine;           /* CMAC cipher engine (legacy) */
+};
+
+static void *cmac_new(void *provctx)
+{
+    struct cmac_data_st *macctx;
+
+    if ((macctx = OPENSSL_zalloc(sizeof(*macctx))) == NULL
+        || (macctx->ctx = CMAC_CTX_new()) == NULL) {
+        OPENSSL_free(macctx);
+        macctx = NULL;
+    }
+    macctx->provctx = provctx;
+
+    return macctx;
+}
+
+static void cmac_free(void *vmacctx)
+{
+    struct cmac_data_st *macctx = vmacctx;
+
+    if (macctx != NULL) {
+        CMAC_CTX_free(macctx->ctx);
+        EVP_CIPHER_meth_free(macctx->alloc_cipher);
+        OPENSSL_free(macctx);
+    }
+}
+
+static void *cmac_dup(void *vsrc)
+{
+    struct cmac_data_st *src = vsrc;
+    struct cmac_data_st *dst = cmac_new(src->provctx);
+
+    if (!CMAC_CTX_copy(dst->ctx, src->ctx)) {
+        cmac_free(dst);
+        return 0;
+    }
+
+    if (src->alloc_cipher != NULL && !EVP_CIPHER_up_ref(src->alloc_cipher)) {
+        cmac_free(dst);
+        return 0;
+    }
+
+    dst->tmpengine = src->tmpengine;
+    dst->tmpcipher = src->tmpcipher;
+    dst->alloc_cipher = src->alloc_cipher;
```

Comment:
Up ref?

Reply:
Done

Reply:
I'm finding more bugs, so there's more to adjust...